crates/pruning/src/lib.rs

```diff
@@ -0,0 +1,109 @@
+
+#[cfg(test)]
+mod tests;
+
+/// A configuration type to define the behaviour of the pruner.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct PruningConfig {
+    /// Defines the number of blocks from the tip of the chain to remove.
+    pub distance: u64,
+}
+
+impl Default for PruningConfig {
+    fn default() -> Self {
+        Self { distance: 256 }
+    }
+}
+
+pub struct Pruner<DB>
+where
+    DB: SharedLedgerOps,
+{
+    /// config
+    config: PruningConfig,
+    /// The last block number which was pruned.
+    last_pruned_block: u64,
+    /// A channel receiver which gets notified of new L2 blocks.
+    l2_receiver: broadcast::Receiver<u64>,
+    /// Access to ledger tables.
+    ledger_db: DB,
+}
+
+impl<DB> Pruner<DB>
+where
+    DB: SharedLedgerOps + Send + Sync + Clone + 'static,
+{
+    pub fn new(
+        config: PruningConfig,
+        last_pruned_block: u64,
+        l2_receiver: broadcast::Receiver<u64>,
+        ledger_db: DB,
+    ) -> Self {
+        Self {
+            config,
+            last_pruned_block,
+            l2_receiver,
+            ledger_db,
+        }
+    }
+
+    /// Prune everything
+    pub async fn prune(&self, up_to_block: u64) {
+        info!("Pruning up to L2 block: {}", up_to_block);
+        let ledger_db = self.ledger_db.clone();
+        let ledger_pruning_handle =
+            tokio::task::spawn_blocking(move || prune_ledger(ledger_db, up_to_block));
+        let evm_pruning_handle = tokio::task::spawn_blocking(move || prune_evm(up_to_block));
+
+        future::join_all([ledger_pruning_handle, evm_pruning_handle]).await;
+    }
+
+    pub async fn run(mut self, cancellation_token: CancellationToken) {
+        loop {
+            select! {
+                biased;
+                _ = cancellation_token.cancelled() => {
+                    // Store the last pruned l2 height in ledger DB to be restored in the next initialization.
+                    if let Err(e) = self.ledger_db.set_last_pruned_l2_height(self.last_pruned_block) {
+                        error!("Failed to store last pruned L2 height {}: {:?}", self.last_pruned_block, e);
+                    }
+                    return;
+                }
+                current_l2_block = self.l2_receiver.recv() => {
+                    if let Ok(current_l2_block) = current_l2_block {
+                        // Calculate the block at which pruning would be triggered.
+                        // This is allowing `self.config.distance` blocks to be produced before we
+                        // decide to prune the previous `self.config.distance` blocks.
+                        let trigger_block = (self.last_pruned_block + self.config.distance) + 1;
+                        if current_l2_block >=trigger_block {
+                            self.prune(current_l2_block - self.config.distance).await;
+                            self.last_pruned_block += self.config.distance;
+                        }
```

Comment:
I think the naming of last_pruned_block is wrong. Assume distance = 1000 scenario:
1. chain produced 1001th block
2. trigger_block = 0 + 1000 + 1 = 1001, hence triggering a prune
3. prune(1001 - 1000) = prune(1), so nothing is pruned assuming up_to_block is exclusive
4. last_pruned_block = 1000
As you see even though we didn't prune anything, last_pruned_block became 1000. And in the next iteration when chain reaches 2001, 1 to 1000 blocks will be pruned, and last_pruned_block will become 2000. I think more appropriate name is something like next_pruning_end_block.

Reply:
Good observation.

I have changed this to:
```rust
self.prune(self.last_pruned_block + self.config.distance).await;
```

So when we start:
- trigger block = 1001
- prune(up_to_block = 1000)
- last_pruned_block = 1000

Next iteration
- last_pruned_block = 1000
- trigger_block = 2001
- prune(up_to_block = 2000)
- last_pruned_block = 2000

... etc

Reply:
Now it is really last_pruned_block, but I am not fond of this pruning algorithm tbh. Because at trigger points, there will only remain 1 block, which is not good imo because while at 1001th block, its not feasible to not be able to query 1000th block, its still too new and most likely useful. I think if the config is set to 1000, we should always have at least 1000 blocks, and always prune down to last 1000 when a certain condition is met.

I propose the condition to be something like 2x of distance limit. So anytime the stored block count in the database reaches 2000, it can be pruned to its latest 1000 blocks.

Reply:
wdyt @eyusufatik @kpp

Reply:
You're totally right. I have not considered the fact that we should keep more than 1 blocks at any given moment. Although i meant to do that given that Reth's and Substrate's implementation do based on my research.

I'll amend

Reply:
Done